### ISSUE TITLE
ERROR: Input file is not bgzipped, cannot use tabix

### ISSUE THREAD

**Issue by moldach** [Ensembl/ensembl-vep#817]
### System
- VEP version: 
```
[moldach@cedar5 VCF-FILES]$ ~/bin/ensembl-vep/vep
#----------------------------------#
# ENSEMBL VARIANT EFFECT PREDICTOR #
#----------------------------------#

Versions:
  ensembl              : 100.171092c
  ensembl-funcgen      : 100.f0c3948
  ensembl-io           : 100.f87ae4f
  ensembl-variation    : 100.b220ff4
  ensembl-vep          : 100.4
```
- Perl version: 5.32.0
- OS: CentOS
- tabix installed yes
```
[moldach@cedar5 VCF-FILES]$ which tabix
/cvmfs/soft.computecanada.ca/easybuild/software/2017/avx2/Compiler/intel2016.4/htslib/1.5/bin/tabix
```

I'm getting an error that VCF files are not bgzipped (even though they are):

Here is a head of the file:

```
[moldach@cedar5 VCF-FILES]$ head -n 30 pindel.vcf
##fileformat=VCFv4.0
##fileDate=00000000
##source=pindel
##reference=WS265.PRJNA13758
##INFO=<ID=END,Number=1,Type=Integer,Description="End position of the variant described in this record">
##INFO=<ID=HOMLEN,Number=1,Type=Integer,Description="Length of base pair identical micro-homology at event breakpoints">
##INFO=<ID=PF,Number=1,Type=Integer,Description="The number of samples carry the variant">
##INFO=<ID=HOMSEQ,Number=.,Type=String,Description="Sequence of base pair identical micro-homology at event breakpoints">
##INFO=<ID=SVLEN,Number=1,Type=Integer,Description="Difference in length between REF and ALT alleles">
##INFO=<ID=SVTYPE,Number=1,Type=String,Description="Type of structural variant">
##INFO=<ID=NTLEN,Number=.,Type=Integer,Description="Number of bases inserted in place of deleted code">
##FORMAT=<ID=PL,Number=3,Type=Integer,Description="Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=RD,Number=1,Type=Integer,Description="Reference depth, how many reads support the reference">
##FORMAT=<ID=AD,Number=2,Type=Integer,Description="Allele depth, how many reads support this allele">
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  maddog_pipeline
II      5008800 .       GA      G       .       PASS    END=5008801;HOMLEN=9;HOMSEQ=AAAAAAAAA;SVLEN=-1;SVTYPE=DEL   GT:AD   0/0:32,2
```

## Method 1:

```
bgzip -c pindel.vcf > pindel.vcf.gz
tabix -p vcf pindel.vcf.gz
```

## Method 2:

```
bcftools view -Oz -o pindel.vcf.gz pindel.vcf
htsfile pindel.vcf.gz
bcftools index pindel.vcf.gz
```

Both give the error with the following command:

Here is the command for `vep`:
```
~/bin/ensembl-vep/vep -i pindel.vcf.gz \
--species "caenorhabditis_elegans" \
--format "vcf" \
--everything \
--database \
--force_overwrite \
--fasta c_elegans.PRJNA13758.WS265.genomic.fa \
--gff c_elegans.PRJNA13758.WS265.annotations.gff3 \
--tab \
--variant_class \
--regulatory \
--show_ref_allele \
--numbers \
--symbol \
--protein \
--pubmed \
--check_existing \
--exclude_null_allele \
-o pindel_annotated.vcf
```

**Comment by sarahhunt** [comment]
Hi @moldach,

That error is a little missleading - it is the GFF3 file that needs to be indexed, rather than the variants you wish to annotate. VEP goes through the variant input file reading a chunk of lines at a time then pulls out features at the relevant coordinates using the index on the GFF. There are details on preparing the file and the information required here:

http://www.ensembl.org/info/docs/tools/vep/script/vep_cache.html#gff

Also, we don't have variant data or citations available in our databases for C.elegans, so check_existing, pubmed and exclude_null_allele will not return data.

**Comment by moldach** [comment]
Hi @sarahhunt thanks for clearing this up and for suggesting to omit those three params.


After indexing the GFF3 I've tried the following:

```
~/bin/ensembl-vep/vep \
        -i pindel.vcf.gz \
        --species "caenorhabditis_elegans" \
        --format "vcf" \
        --everything \
        --database \
        --force_overwrite \
        --fasta c_elegans.PRJNA13758.WS265.genomic.fa \
        --gff c_elegans.PRJNA13758.WS265.annotations.gff3.gz \
        --tab \
        --variant_class \
        --regulatory \
        --show_ref_allele \
        --numbers \
        --symbol \
        --protein \
        -o pindel_annotated.vcf
```

I got an e-mail from the SLURM job scheduler after **4 hours and 23 minutes** that the job **FAILED** with ExitCode 1

In the `.err` file I see:

```
gzip: stdout: Broken pipe
Use of uninitialized value in concatenation (.) or string at /project/6013424/common/tools/ensembl-vep/modules/Bio/EnsEMBL/VEP/OutputFactory/BaseTab.pm line 101.

gzip: stdout: Broken pipe
```

Any idea what went wrong or how to troubleshoot this?

**Comment by at7** [comment]
Hi @moldach,

we do see the `Broken pipe` error occasionally but don't really have an explanation for how the error is caused. The other message is more like a warning message and shouldn't be the failure reason.
Could you please check your output file? I wonder if the `Broken pipe` message causes your job scheduler to return a  non-zero code despite VEP completing the annotations successfully.
I'm very sorry that I cannot give you a better reply and would also offer to take a look at your input data to see if I can reproduce the problem.

Thank you,
Anja

**Comment by moldach** [comment]
The output file only has the header lines but not annotations.

I would appreciate if you could try to reproduce the problem on your end.

I'm sending you an e-mail to your ebi account with instructions for downloading the references (fa/gff3) and a dropbox link to my `Pindel` VCF file.

**Comment by at7** [comment]
Thank you for sending all the input files. I have downloaded them and will start debugging the problem you described now. I get back to you when I know more.

**Comment by at7** [comment]
I tried to index your GFF3 file and got an error: Not a BGZF file: c_elegans.PRJNA13758.WS265.annotations.gff3.gz.
I then uncompressed and ran the following commands:
grep -v "#" data.gff | sort -k1,1 -k4,4n -k5,5n -t$'\t' | bgzip -c > data.gff.gz
tabix -p gff data.gff.gz
The steps are described here: https://www.ensembl.org/info/docs/tools/vep/script/vep_cache.html#gff

After that I was able to run VEP and get annotations with your VEP command line and input data.

I also noticed a few warning messages related to your GFF3 file:
  WARNING: Unable to determine biotype of ..
  WARNING: Parent entries with the following IDs were not found or skipped due to invalid types ..
We describe VEP's GFF format expectations here as well: http://www.ensembl.org/info/docs/tools/vep/script/vep_cache.html#gff

And finally, using the --database option instructs VEP to also look up transcripts from the Ensembl database which for release 100 come from assembly version WBcel235 for caenorhabditis elegans.

To summarise, after recompressing and indexing the GFF3 file I didn't have problems to run VEP for your input data. If you don't want VEP to use the annotations from the Ensembl database just swap the --database option with --offline.

Could you please let me know how you compressed and indexed your GFF3 file and if following the instructions from our documentation page fixes the problem?

Thank you,
Anja

**Comment by moldach** [comment]
Hi @at7 thank you very much for all of that information.

The only thing I can surmise is that there was likely a segmentation fault (OOM) when I was running `grep -v "#" data.gff | sort -k1,1 -k4,4n -k5,5n -t$'\t' | bgzip -c > data.gff.gz` interactively which resulted in a truncated file and that's why it wouldn't index.

Running these steps again has solved the issue.

**Comment by moldach** [comment]
Could you please clear up the following:

> For optimum performance, download a cache file for your species of interest, using either the installer or by following the VEP Cache documentation, and run VEP **with either the `--cache` or `--offline` option.**

What is the difference between using `--cache` or `--offline`?

**Comment by moldach** [comment]
> I also noticed a few warning messages related to your GFF3 file:
WARNING: Unable to determine biotype of ..
WARNING: Parent entries with the following IDs were not found or skipped due to invalid types ..
We describe VEP's GFF format expectations here as well: http://www.ensembl.org/info/docs/tools/vep/script/vep_cache.html#gff

The [reference genome and annotation (GFF3) are from WormBase](ftp://ftp.wormbase.org/pub/wormbase/releases/WS265/species/c_elegans/PRJNA13758/) so they refer to the same genome. 

As for format expectations:

> Due to inconsistency in the GFF specification and adherence to it, VEP may encounter problems parsing some GFF files. For the same reason, not all transcript biotypes defined in your GFF may be supported by VEP.

Is this something I should bring up the curators of the GFF3 (WormBase)?
How serious of a concern is this, and, is there anything _I can do_?

**Comment by at7** [comment]
With regard to your question about the --offline and --cache options:
If you only use --cache VEP will get all transcript information from the cache files. VEP will still use a database connection to fetch data that is not in the cache like fetching reference sequence for HGVS calculations. If you also use --offline VEP won't get the reference sequence from the database. In which case you also need to provide a FAST file with the reference sequence using the --fasta option. Only using --offline will automatically enable the --cache option.

With regard to the GFF3 file: The GFF file will still work, but it'll likely be missing some data. If the type column doesn't contain one of these values: https://www.ensembl.org/info/docs/tools/vep/script/vep_cache.html#gfftypes then VEP pretends the line doesn't exist. Similarly, if there's a reference to a parent that doesn't exist, then that attribute is skipped as well. Please contact WormBase in order to find out which entries in the GFF3 file are affected by the missing information to help you better understand which annotations you might be missing.

**Comment by nuno-agostinho** [comment]
Hey @moldach, hope you are having a nice day!

I would like to know if you still need help regarding this topic.

Sorry for any inconvenience.

Thanks,
Nuno

**Comment by moldach** [comment]
I no longer work for this client so I've closed the issue @nuno-agostinho
